diff --git a/.gitmodules b/.gitmodules
index c6b1cb9c..7acbd4e9 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,15 +1,15 @@
 [submodule "ip_cores/general-cores"]
 	path = ip_cores/general-cores
-	url = https://ohwr.org/project/general-cores.git
+	url = https://gitlab.com/ohwr/project/general-cores.git
 [submodule "ip_cores/etherbone-core"]
 	path = ip_cores/etherbone-core
-	url = https://ohwr.org/project/etherbone-core.git
+	url = https://gitlab.com/ohwr/project/etherbone-core.git
 [submodule "ip_cores/gn4124-core"]
 	path = ip_cores/gn4124-core
-	url = https://ohwr.org/project/gn4124-core.git
+	url = https://gitlab.com/ohwr/project/gn4124-core.git
 [submodule "ip_cores/vme64x-core"]
 	path = ip_cores/vme64x-core
-	url = https://ohwr.org/project/vme64x-core.git
+	url = https://gitlab.com/ohwr/project/vme64x-core.git
 [submodule "ip_cores/urv-core"]
 	path = ip_cores/urv-core
-	url = https://ohwr.org/project/urv-core.git
+	url = https://gitlab.com/ohwr/project/urv-core.git
diff --git a/board/Manifest.py b/board/Manifest.py
index 41c8c8eb..bdede8c2 100644
--- a/board/Manifest.py
+++ b/board/Manifest.py
@@ -1,5 +1,5 @@
 try:
-    if board in ["spec", "svec", "vfchd", "clbv2", "clbv3", "clbv4", "pxie-fmc", "diot-sb", "fasec", "zcu10x", "zc706", "common"]:
+    if board in ["m2sdr", "spec", "svec", "vfchd", "clbv2", "clbv3", "clbv4", "pxie-fmc", "diot-sb", "fasec", "zcu10x", "zc706", "common"]:
         modules = {"local" : [ board ] }
 except NameError:
     # board is not defined
diff --git a/ip_cores/general-cores b/ip_cores/general-cores
--- a/ip_cores/general-cores
+++ b/ip_cores/general-cores
@@ -1 +1 @@
-Subproject commit dc8d765c669c35b0d396b5afe5085c08686f7330
+Subproject commit dc8d765c669c35b0d396b5afe5085c08686f7330-dirty
diff --git a/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper.vhd b/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper.vhd
index f981871f..420aacdd 100644
--- a/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper.vhd
+++ b/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper.vhd
@@ -75,7 +75,8 @@ generic
 (
     -- Simulation attributes
     EXAMPLE_SIMULATION             : integer  := 0;      -- Set to 1 for simulation
-    WRAPPER_SIM_GTRESET_SPEEDUP    : string   := "FALSE" -- Set to "true" to speed up sim reset
+    WRAPPER_SIM_GTRESET_SPEEDUP    : string   := "FALSE"; -- Set to "true" to speed up sim reset
+    g_phy_refclk_sel               : integer  := 1
 );
 port
 (
@@ -109,6 +110,7 @@ port
     ------------------------ Receive Ports - RX AFE Ports ----------------------
     GT0_GTPRXN_IN                           : in   std_logic;
     GT0_GTPRXP_IN                           : in   std_logic;
+    RX_POLARITY                              : in   std_logic;
     -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
     GT0_RXBYTEISALIGNED_OUT                 : out  std_logic;
     GT0_RXCOMMADET_OUT                      : out  std_logic;
@@ -134,6 +136,7 @@ port
     --------------- Transmit Ports - TX Configurable Driver Ports --------------
     GT0_GTPTXN_OUT                          : out  std_logic;
     GT0_GTPTXP_OUT                          : out  std_logic;
+    TX_POLARITY                              : in   std_logic;
     ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
     GT0_TXOUTCLK_OUT                        : out  std_logic;
     GT0_TXOUTCLKFABRIC_OUT                  : out  std_logic;
@@ -143,7 +146,7 @@ port
     ------------------ Transmit Ports - pattern Generator Ports ----------------
     GT0_TXPRBSSEL_IN                        : in   std_logic_vector(2 downto 0);
 
-	
+
     --____________________________COMMON PORTS________________________________
     ----------------- Common Block - GTPE2_COMMON Clocking Ports ---------------
     GT0_GTREFCLK0_IN                        : in   std_logic;
@@ -190,7 +193,10 @@ architecture RTL of whiterabbit_gtpe2_channel_wrapper is
     signal   gt0_pll1clk_i       :   std_logic;
     signal   gt0_pll1refclk_i    :   std_logic;
     signal    gt0_rst_i                       : std_logic;
-      
+
+    signal gtp_common_refclk_ext : std_logic;
+    signal gtp_common_refclk_internal : std_logic;
+    signal gtp_common_refclksel : std_logic_vector(2 downto 0);
 
 
 
@@ -200,15 +206,16 @@ generic
 (
     -- Simulation attributes
     GT_SIM_GTRESET_SPEEDUP    : string := "false";
-    EXAMPLE_SIMULATION        : integer  := 0;  
+    EXAMPLE_SIMULATION        : integer  := 0;
     TXSYNC_OVRD_IN            : bit    := '0';
-    TXSYNC_MULTILANE_IN       : bit    := '0'     
+    TXSYNC_MULTILANE_IN       : bit    := '0';
+    g_phy_refclk_sel          : integer := 1
 );
-port 
-(   
+port
+(
     RST_IN                                  : in   std_logic;
     DRP_BUSY_OUT                            : out  std_logic;
-   
+
     ---------------------------- Channel - DRP Ports  --------------------------
     DRPADDR_IN                              : in   std_logic_vector(8 downto 0);
     DRPCLK_IN                               : in   std_logic;
@@ -240,6 +247,7 @@ port
     ------------------------ Receive Ports - RX AFE Ports ----------------------
     GTPRXN_IN                               : in   std_logic;
     GTPRXP_IN                               : in   std_logic;
+    RX_POLARITY                              : in   std_logic;
     -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
     RXBYTEISALIGNED_OUT                     : out  std_logic;
     RXCOMMADET_OUT                          : out  std_logic;
@@ -265,6 +273,7 @@ port
     --------------- Transmit Ports - TX Configurable Driver Ports --------------
     GTPTXN_OUT                              : out  std_logic;
     GTPTXP_OUT                              : out  std_logic;
+    TX_POLARITY                              : in   std_logic;
     ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
     TXOUTCLK_OUT                            : out  std_logic;
     TXOUTCLKFABRIC_OUT                      : out  std_logic;
@@ -288,20 +297,20 @@ end component;
 
 --********************************* Main Body of Code**************************
 
-begin                       
+begin
 
     tied_to_ground_i                    <= '0';
     tied_to_ground_vec_i(63 downto 0)   <= (others => '0');
     tied_to_vcc_i                       <= '1';
-    gt0_pll0clk_i    <= gt0_pll0outclk_i;  
-    gt0_pll0refclk_i <= gt0_pll0outrefclk_i; 
-    gt0_pll1clk_i    <= gt0_pll1outclk_i;  
-    gt0_pll1refclk_i <= gt0_pll1outrefclk_i; 
-      
+    gt0_pll0clk_i    <= gt0_pll0outclk_i;
+    gt0_pll0refclk_i <= gt0_pll0outrefclk_i;
+    gt0_pll1clk_i    <= gt0_pll1outclk_i;
+    gt0_pll1refclk_i <= gt0_pll1outrefclk_i;
+
     gt0_rst_i        <= GT0_PLL1RESET_IN;
-   
-    
-    --------------------------- GT Instances  -------------------------------   
+
+
+    --------------------------- GT Instances  -------------------------------
     --_________________________________________________________________________
     --_________________________________________________________________________
     GT_INST : whiterabbit_gtpe2_channel_wrapper_gt
@@ -349,6 +358,7 @@ begin
         ------------------------ Receive Ports - RX AFE Ports ----------------------
         GTPRXN_IN                       =>      GT0_GTPRXN_IN,
         GTPRXP_IN                       =>      GT0_GTPRXP_IN,
+        RX_POLARITY                     =>      RX_POLARITY,
         -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
         RXBYTEISALIGNED_OUT             =>      GT0_RXBYTEISALIGNED_OUT,
         RXCOMMADET_OUT                  =>      GT0_RXCOMMADET_OUT,
@@ -374,6 +384,7 @@ begin
         --------------- Transmit Ports - TX Configurable Driver Ports --------------
         GTPTXN_OUT                      =>      GT0_GTPTXN_OUT,
         GTPTXP_OUT                      =>      GT0_GTPTXP_OUT,
+        TX_POLARITY                     =>      TX_POLARITY,
         ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
         TXOUTCLK_OUT                    =>      GT0_TXOUTCLK_OUT,
         TXOUTCLKFABRIC_OUT              =>      GT0_TXOUTCLKFABRIC_OUT,
@@ -390,6 +401,17 @@ begin
     --_________________________________________________________________________
     --_________________________GTPE2_COMMON____________________________________
 
+    gen_gtp_refclk_internal : if (g_phy_refclk_sel = 7) generate
+      gtp_common_refclk_internal <= GT0_GTREFCLK0_IN;
+      gtp_common_refclk_ext <= tied_to_ground_i;
+      gtp_common_refclksel <= "111";
+    end generate gen_gtp_refclk_internal;
+    gen_gtp_refclk_external : if (g_phy_refclk_sel /= 7) generate
+      gtp_common_refclk_internal <= tied_to_ground_i;
+      gtp_common_refclk_ext <= GT0_GTREFCLK0_IN;
+      gtp_common_refclksel <= "001";
+    end generate gen_gtp_refclk_external;
+
     gtpe2_common_0_i : GTPE2_COMMON
     generic map
     (
@@ -399,12 +421,12 @@ begin
             SIM_PLL1REFCLK_SEL   => ("001"),
             SIM_VERSION          => ("1.0"),
 
-            PLL0_FBDIV           => PLL0_FBDIV_IN     ,	
-	    PLL0_FBDIV_45        => PLL0_FBDIV_45_IN  ,	
-	    PLL0_REFCLK_DIV      => PLL0_REFCLK_DIV_IN,	
-	    PLL1_FBDIV           => PLL1_FBDIV_IN     ,	
-	    PLL1_FBDIV_45        => PLL1_FBDIV_45_IN  ,	
-	    PLL1_REFCLK_DIV      => PLL1_REFCLK_DIV_IN,	            
+            PLL0_FBDIV           => PLL0_FBDIV_IN     ,
+	    PLL0_FBDIV_45        => PLL0_FBDIV_45_IN  ,
+	    PLL0_REFCLK_DIV      => PLL0_REFCLK_DIV_IN,
+	    PLL1_FBDIV           => PLL1_FBDIV_IN     ,
+	    PLL1_FBDIV_45        => PLL1_FBDIV_45_IN  ,
+	    PLL1_REFCLK_DIV      => PLL1_REFCLK_DIV_IN,
 
 
        ------------------COMMON BLOCK Attributes---------------
@@ -426,7 +448,6 @@ begin
         RSVD_ATTR0                              =>     (x"0000"),
         RSVD_ATTR1                              =>     (x"0000")
 
-        
     )
     port map
     (
@@ -442,8 +463,9 @@ begin
         ----------------- Common Block - GTPE2_COMMON Clocking Ports ---------------
         GTEASTREFCLK0                   =>      tied_to_ground_i,
         GTEASTREFCLK1                   =>      tied_to_ground_i,
-        GTGREFCLK1                      =>      tied_to_ground_i,
-        GTREFCLK0                       =>      GT0_GTREFCLK0_IN,
+        GTGREFCLK0                      =>      gtp_common_refclk_internal, -- clk input via internal fpga path
+        GTGREFCLK1                      =>      gtp_common_refclk_internal,
+        GTREFCLK0                       =>      gtp_common_refclk_ext, -- clk input via dedicated pins
         GTREFCLK1                       =>      tied_to_ground_i,
         GTWESTREFCLK0                   =>      tied_to_ground_i,
         GTWESTREFCLK1                   =>      tied_to_ground_i,
@@ -458,7 +480,7 @@ begin
         PLL0LOCKEN                      =>      tied_to_vcc_i,
         PLL0PD                          =>      '1',
         PLL0REFCLKLOST                  =>      open,
-        PLL0REFCLKSEL                   =>      "001",
+        PLL0REFCLKSEL                   =>      gtp_common_refclksel,
         PLL0RESET                       =>      tied_to_ground_i,
         PLL1FBCLKLOST                   =>      open,
         PLL1LOCK                        =>      GT0_PLL1LOCK_OUT,
@@ -466,11 +488,10 @@ begin
         PLL1LOCKEN                      =>      tied_to_vcc_i,
         PLL1PD                          =>      tied_to_ground_i,
         PLL1REFCLKLOST                  =>      GT0_PLL1REFCLKLOST_OUT,
-        PLL1REFCLKSEL                   =>      "001",
+        PLL1REFCLKSEL                   =>      gtp_common_refclksel,
         PLL1RESET                       =>      GT0_PLL1RESET_IN,
         ---------------------------- Common Block - Ports --------------------------
         BGRCALOVRDENB                   =>      tied_to_vcc_i,
-        GTGREFCLK0                      =>      tied_to_ground_i,
         PLLRSVD1                        =>      "0000000000000000",
         PLLRSVD2                        =>      "00000",
         REFCLKOUTMONITOR0               =>      open,
diff --git a/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper_gt.vhd b/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper_gt.vhd
index 90ca8f39..645589ee 100644
--- a/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper_gt.vhd
+++ b/platform/xilinx/wr_gtp_phy/family7-gtp/whiterabbit_gtpe2_channel_wrapper_gt.vhd
@@ -113,6 +113,7 @@ port
     ------------------------ Receive Ports - RX AFE Ports ----------------------
     gtprxn_in                               : in   std_logic;
     gtprxp_in                               : in   std_logic;
+    rx_polarity                             : in  std_logic             := '0';
     -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
     rxbyteisaligned_out                     : out  std_logic;
     rxcommadet_out                          : out  std_logic;
@@ -138,6 +139,7 @@ port
     --------------- Transmit Ports - TX Configurable Driver Ports --------------
     gtptxn_out                              : out  std_logic;
     gtptxp_out                              : out  std_logic;
+    tx_polarity                             : in  std_logic             := '0';
     ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
     txoutclk_out                            : out  std_logic;
     txoutclkfabric_out                      : out  std_logic;
@@ -810,7 +812,7 @@ begin
         RXELECIDLE                      =>      open,
         RXELECIDLEMODE                  =>      "11",
         ----------------- Receive Ports - RX Polarity Control Ports ----------------
-        RXPOLARITY                      =>      tied_to_ground_i,
+        RXPOLARITY                      =>      rx_polarity,
         -------------- Receive Ports -RX Initialization and Reset Ports ------------
         RXRESETDONE                     =>      RXRESETDONE_OUT,
         --------------------------- TX Buffer Bypass Ports -------------------------
@@ -914,7 +916,7 @@ begin
         TXCOMWAKE                       =>      tied_to_ground_i,
         TXPDELECIDLEMODE                =>      tied_to_ground_i,
         ----------------- Transmit Ports - TX Polarity Control Ports ---------------
-        TXPOLARITY                      =>      tied_to_ground_i,
+        TXPOLARITY                      =>      tx_polarity,
         --------------- Transmit Ports - TX Receiver Detection Ports  --------------
         TXDETECTRX                      =>      tied_to_ground_i,
         ------------------ Transmit Ports - pattern Generator Ports ----------------
diff --git a/platform/xilinx/wr_gtp_phy/family7-gtp/wr_gtp_phy_family7.vhd b/platform/xilinx/wr_gtp_phy/family7-gtp/wr_gtp_phy_family7.vhd
index 4f6a8fe6..11f58313 100644
--- a/platform/xilinx/wr_gtp_phy/family7-gtp/wr_gtp_phy_family7.vhd
+++ b/platform/xilinx/wr_gtp_phy/family7-gtp/wr_gtp_phy_family7.vhd
@@ -55,7 +55,8 @@ use work.disparity_gen_pkg.all;
 entity wr_gtp_phy_family7 is
   generic (
     -- set to non-zero value to speed up the simulation by reducing some delays
-    g_simulation     : integer := 0
+    g_simulation     : integer := 0;
+    g_phy_refclk_sel : integer := 1
   );
   port (
     -- Dedicated reference 125 MHz clock for the GTP transceiver
@@ -110,6 +111,9 @@ entity wr_gtp_phy_family7 is
     pad_txn_o        : out  std_logic;
     pad_txp_o        : out  std_logic;
 
+    tx_polarity  : in  std_logic             := '0';
+    rx_polarity  : in  std_logic             := '0';
+
     rdy_o            : out  std_logic
   );
 end entity wr_gtp_phy_family7;
@@ -172,7 +176,8 @@ architecture structure of wr_gtp_phy_family7 is
   (
       -- Simulation attributes
       EXAMPLE_SIMULATION             : integer  := 0;      -- Set to 1 for simulation
-      WRAPPER_SIM_GTRESET_SPEEDUP    : string := "FALSE" -- Set to "true" to speed up sim reset
+      WRAPPER_SIM_GTRESET_SPEEDUP    : string := "FALSE"; -- Set to "true" to speed up sim reset
+      g_phy_refclk_sel               : integer := 1
   );
   port
   (
@@ -205,6 +210,7 @@ architecture structure of wr_gtp_phy_family7 is
       ------------------------ Receive Ports - RX AFE Ports ----------------------
       GT0_GTPRXN_IN            : in  std_logic;
       GT0_GTPRXP_IN            : in  std_logic;
+      RX_POLARITY              : in  std_logic;
       -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
       GT0_RXBYTEISALIGNED_OUT  : out  std_logic;
       GT0_RXCOMMADET_OUT       : out  std_logic;
@@ -230,6 +236,7 @@ architecture structure of wr_gtp_phy_family7 is
       --------------- Transmit Ports - TX Configurable Driver Ports --------------
       GT0_GTPTXN_OUT           : out std_logic;
       GT0_GTPTXP_OUT           : out std_logic;
+      TX_POLARITY              : in  std_logic;
       ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
       GT0_TXOUTCLK_OUT         : out std_logic;
       GT0_TXOUTCLKFABRIC_OUT   : out std_logic;
@@ -385,7 +392,8 @@ begin
   (
     -- Simulation attributes
     EXAMPLE_SIMULATION           => g_simulation,
-    WRAPPER_SIM_GTRESET_SPEEDUP  => f_to_bool(g_simulation)
+    WRAPPER_SIM_GTRESET_SPEEDUP  => f_to_bool(g_simulation),
+    g_phy_refclk_sel             => g_phy_refclk_sel
   )
   port map
   (
@@ -419,6 +427,7 @@ begin
     ------------------------ Receive Ports - RX AFE Ports ----------------------
     GT0_GTPRXN_IN            =>  pad_rxn_i,
     GT0_GTPRXP_IN            =>  pad_rxp_i,
+    RX_POLARITY              =>  rx_polarity,
     -------------- Receive Ports - RX Byte and Word Alignment Ports ------------
     GT0_RXBYTEISALIGNED_OUT  =>  rx_byte_is_aligned,
     GT0_RXCOMMADET_OUT       =>  rx_comma_det,
@@ -444,6 +453,7 @@ begin
     --------------- Transmit Ports - TX Configurable Driver Ports --------------
     GT0_GTPTXN_OUT           =>  pad_txn_o,
     GT0_GTPTXP_OUT           =>  pad_txp_o,
+    TX_POLARITY              =>  tx_polarity,
     ----------- Transmit Ports - TX Fabric Clock Output Control Ports ----------
     GT0_TXOUTCLK_OUT         =>  tx_out_clk_bufin,
     GT0_TXOUTCLKFABRIC_OUT   =>  open,
diff --git a/platform/xilinx/wr_xilinx_pkg.vhd b/platform/xilinx/wr_xilinx_pkg.vhd
index 1467311d..43184166 100644
--- a/platform/xilinx/wr_xilinx_pkg.vhd
+++ b/platform/xilinx/wr_xilinx_pkg.vhd
@@ -97,6 +97,8 @@ package wr_xilinx_pkg is
       sfp_tx_fault_i        : in  std_logic             := '0';
       sfp_los_i             : in  std_logic             := '0';
       sfp_tx_disable_o      : out std_logic;
+      sfp_tx_polarity       : in  std_logic             := '0';  
+      sfp_rx_polarity       : in  std_logic             := '0';  
       sfp_mux_sel_i         : in  std_logic              := '0';
       sfp1_txn_o            : out std_logic;
       sfp1_txp_o            : out std_logic;
@@ -285,7 +287,8 @@ package wr_xilinx_pkg is
   component wr_gtp_phy_family7 is
     generic (
       -- set to non-zero value to speed up the simulation by reducing some delays
-      g_simulation : integer := 0);
+      g_simulation     : integer := 0;
+      g_phy_refclk_sel : integer := 1);
     port (
       clk_gtp_i      : in  std_logic;
       tx_out_clk_o   : out std_logic;
@@ -306,6 +309,8 @@ package wr_xilinx_pkg is
       pad_txp_o      : out std_logic;
       pad_rxn_i      : in  std_logic := '0';
       pad_rxp_i      : in  std_logic := '0';
+      tx_polarity       : in  std_logic             := '0';
+      rx_polarity       : in  std_logic             := '0';
       rdy_o          : out std_logic);
   end component;
 
diff --git a/platform/xilinx/xwrc_platform_vivado.vhd b/platform/xilinx/xwrc_platform_vivado.vhd
index f9943cb9..de7f1829 100644
--- a/platform/xilinx/xwrc_platform_vivado.vhd
+++ b/platform/xilinx/xwrc_platform_vivado.vhd
@@ -140,6 +140,9 @@ entity xwrc_platform_xilinx is
     sfp_tx_fault_i        : in  std_logic             := '0';
     sfp_los_i             : in  std_logic             := '0';
     sfp_tx_disable_o      : out std_logic;
+    sfp_tx_polarity       : in  std_logic             := '0';  
+    sfp_rx_polarity       : in  std_logic             := '0';  
+
     ---------------------------------------------------------------------------
     -- if both SFP channels are enabled and sfp_mux is enabled,
     -- this is the bit to switch between them
@@ -904,17 +907,23 @@ begin  -- architecture rtl
   begin
 
     -- Dedicated GTP clock.
-    cmp_gtp_dedicated_clk : IBUFDS_GTE2
-      generic map(
-        CLKCM_CFG    => true,
-        CLKRCV_TRST  => true,
-        CLKSWING_CFG => "11")
-      port map (
-        O     => clk_125m_gtp_buf,
-        ODIV2 => open,
-        CEB   => '0',
-        I     => clk_125m_gtp_p_i,
-        IB    => clk_125m_gtp_n_i);
+    -- g_phy_refclk_sel = 7 is FPGA internal path, clk_125m_gtp_p_i has to be a clocking signal and clk_125m_gtp_n_i is discarded
+    gen_gtp_dedicated_clk_internal : if(g_phy_refclk_sel = 7) generate
+      clk_125m_gtp_buf <= clk_125m_gtp_p_i;
+    end generate gen_gtp_dedicated_clk_internal;
+    gen_gtp_dedicated_clk_external : if(g_phy_refclk_sel /= 7) generate
+      cmp_gtp_dedicated_clk : IBUFDS_GTE2
+        generic map(
+          CLKCM_CFG    => true,
+          CLKRCV_TRST  => true,
+          CLKSWING_CFG => "11")
+        port map (
+          O     => clk_125m_gtp_buf,
+          ODIV2 => open,
+          CEB   => '0',
+          I     => clk_125m_gtp_p_i,
+          IB    => clk_125m_gtp_n_i);
+    end generate gen_gtp_dedicated_clk_external;
 
     -- System PLL input clock buffer
     cmp_clk_sys_buf_i : BUFG
@@ -924,7 +933,8 @@ begin  -- architecture rtl
 
     cmp_gtp: wr_gtp_phy_family7
       generic map(
-        g_simulation => g_simulation)
+        g_simulation => g_simulation,
+        g_phy_refclk_sel => g_phy_refclk_sel)
       port map(
         clk_gtp_i      => clk_125m_gtp_buf,
         tx_out_clk_o   => clk_ref,
@@ -947,6 +957,9 @@ begin  -- architecture rtl
         pad_rxn_i => sfp_rxn_i,
         pad_rxp_i => sfp_rxp_i,
 
+        tx_polarity => sfp_tx_polarity,
+        rx_polarity => sfp_rx_polarity,
+
         tx_locked_o   => clk_ref_locked);
 
     clk_125m_ref_o       <= clk_ref;
diff --git a/platform/xilinx/xwrc_platform_xilinx.vhd b/platform/xilinx/xwrc_platform_xilinx.vhd
index 24db6ccf..77a3cb26 100644
--- a/platform/xilinx/xwrc_platform_xilinx.vhd
+++ b/platform/xilinx/xwrc_platform_xilinx.vhd
@@ -128,6 +128,9 @@ entity xwrc_platform_xilinx is
     sfp_tx_fault_i        : in  std_logic             := '0';
     sfp_los_i             : in  std_logic             := '0';
     sfp_tx_disable_o      : out std_logic;
+    sfp_tx_polarity       : in  std_logic             := '0';  
+    sfp_rx_polarity       : in  std_logic             := '0';  
+
     ---------------------------------------------------------------------------
     -- if both SFP channels are enabled and sfp_mux is enabled, 
     -- this is the bit to switch between them
@@ -1253,17 +1256,23 @@ begin  -- architecture rtl
   begin
 
     -- Dedicated GTP clock.
-    cmp_gtp_dedicated_clk : IBUFDS_GTE2
-      generic map(
-        CLKCM_CFG    => true,
-        CLKRCV_TRST  => true,
-        CLKSWING_CFG => "11")
-      port map (
-        O     => clk_125m_gtp_buf,
-        ODIV2 => open,
-        CEB   => '0',
-        I     => clk_125m_gtp_p_i,
-        IB    => clk_125m_gtp_n_i);
+    -- g_phy_refclk_sel = 7 is FPGA internal path, clk_125m_gtp_p_i has to be a clocking signal and clk_125m_gtp_n_i is discarded
+    gen_gtp_dedicated_clk_internal : if(g_phy_refclk_sel = 7) generate
+      clk_125m_gtp_buf <= clk_125m_gtp_p_i;
+    end generate gen_gtp_dedicated_clk_internal;
+    gen_gtp_dedicated_clk_external : if(g_phy_refclk_sel /= 7) generate
+      cmp_gtp_dedicated_clk : IBUFDS_GTE2
+        generic map(
+          CLKCM_CFG    => true,
+          CLKRCV_TRST  => true,
+          CLKSWING_CFG => "11")
+        port map (
+          O     => clk_125m_gtp_buf,
+          ODIV2 => open,
+          CEB   => '0',
+          I     => clk_125m_gtp_p_i,
+          IB    => clk_125m_gtp_n_i);
+    end generate gen_gtp_dedicated_clk_external;
 
     -- System PLL input clock buffer
     cmp_clk_sys_buf_i : BUFG
@@ -1273,7 +1282,8 @@ begin  -- architecture rtl
 
     cmp_gtp: wr_gtp_phy_family7
       generic map(
-        g_simulation => g_simulation)
+        g_simulation => g_simulation,
+        g_phy_refclk_sel => g_phy_refclk_sel)
       port map(
         clk_gtp_i      => clk_125m_gtp_buf,
         tx_out_clk_o   => clk_ref,
@@ -1296,6 +1306,9 @@ begin  -- architecture rtl
         pad_rxn_i => sfp_rxn_i,
         pad_rxp_i => sfp_rxp_i,
 
+        tx_polarity => sfp_tx_polarity,
+        rx_polarity => sfp_rx_polarity,
+
         tx_locked_o   => clk_ref_locked);
 
     clk_125m_ref_o       <= clk_ref;
